py/server/deephaven/table_listener.py

```diff
@@ -4,15 +4,113 @@
 """ This module provides utilities for listening to table changes. """
 from __future__ import annotations
 
+import inspect
 from abc import ABC, abstractmethod
+from functools import wraps
 from inspect import signature
-from typing import Callable, Union, Type
+from typing import Callable, Union, Type, Sequence, List
 
 import jpy
 
 from deephaven import DHError
+from deephaven._wrapper import JObjectWrapper
+from deephaven.column import Column
+from deephaven.jcompat import to_sequence
+from deephaven.numpy import column_to_numpy_array
 from deephaven.table import Table
 
+_JPythonListenerAdapter = jpy.get_type("io.deephaven.integrations.python.PythonListenerAdapter")
+_JPythonReplayListenerAdapter = jpy.get_type("io.deephaven.integrations.python.PythonReplayListenerAdapter")
+_JTableUpdate = jpy.get_type("io.deephaven.engine.table.TableUpdate")
+_JTableUpdateDataReader = jpy.get_type("io.deephaven.integrations.python.PythonListenerTableUpdateDataReader")
+DEFAULT_CHUNK_SIZE = 4096
+
+
+def _changes_to_numpy(table: Table, col_defs: List[Column], row_set, chunk_size: int, prev: bool = False):
+    row_sequence_iterator = row_set.getRowSequenceIterator()
+    col_sources = [table.j_table.getColumnSource(col_def.name) for col_def in col_defs]
+
+    j_reader_context = _JTableUpdateDataReader.makeContext(chunk_size, *col_sources)
+    while row_sequence_iterator.hasMore():
+        chunk_row_set = row_sequence_iterator.getNextRowSequenceWithLength(chunk_size)
+
+        j_array = _JTableUpdateDataReader.readChunkColumnMajor(j_reader_context, chunk_row_set, col_sources, prev)
+
+        col_dict = {}
+        for i, col_def in enumerate(col_defs):
+            np_array = column_to_numpy_array(col_def, j_array[i])
+            col_dict[col_def.name] = np_array
+
+        yield col_dict
+
+    j_reader_context.close()
+    row_sequence_iterator.close()
+
+
+def _col_defs(table: Table, cols: Union[str, List[str]]):
+    if not cols:
+        col_defs = table.columns
+    else:
+        cols = to_sequence(cols)
+        col_defs = [col for col in table.columns if col.name in cols]
+
+    return col_defs
+
+
+class TableUpdate(JObjectWrapper):
+    j_object_type = _JTableUpdate
+
+    def __init__(self, table: Table, j_table_update):
+        self.table = table
+        self.j_table_update = j_table_update
+        self.chunk_size = DEFAULT_CHUNK_SIZE
+
+    @property
+    def j_object(self) -> jpy.JType:
+        return self.j_table_update
+
+    def added(self, cols: Union[str, List[str]] = None):
```

Comment:
Should there be a convenience argument to return everything as one numpy array, instead of a bunch of split up numpy arrays?